### ISSUE TITLE
C# client's RpcException doesn't have  the "Metadata trailers" property

### ISSUE THREAD

**Issue by RongieZeng** [grpc/grpc#8042]
While the Java Server returns the business errorcode through the "Metadata trailers" in StatusRuntimeException , the C# client can't get the trailers info

```
    public class StatusRuntimeException extends RuntimeException {

      private static final long serialVersionUID = 1950934672280720624L;
      private final Status status;
      private final Metadata trailers;

      public StatusRuntimeException(Status status) {
        this(status, null);
      }

      /**
       * Constructs the exception with both a status and trailers.
       */
      @ExperimentalApi
      public StatusRuntimeException(Status status, @Nullable Metadata trailers) {
        super(Status.formatThrowableMessage(status), status.getCause());
        this.status = status;
        this.trailers = trailers;
      }

      public final Status getStatus() {
        return status;
      }

      @ExperimentalApi
      public final Metadata getTrailers() {
        return trailers;
      }
    }
```

**Comment by jtattermusch** [comment]
You actually can obtain the trailers using the method `GetTrailers()` from the "Call" object.
https://github.com/grpc/grpc/blob/1b5f057995fb2b53746db281ae519d9ae475d24c/src/csharp/Grpc.Core/AsyncUnaryCall.cs#L104

**Comment by RongieZeng** [comment]
But we are using "BlockingUnaryCall", and it seems in the "HandleUnaryResponse" method,  the field "responseHeadersTcs"  which reserve the metadata isn't return, maybe we can add the trailers property to RpcException, and when "status.StatusCode != StatusCode.OK",  set "trailers = responseHeaders"?
`        private void HandleUnaryResponse(bool success, ClientSideStatus receivedStatus, byte[] receivedMessage, Metadata responseHeaders)
        {
            // NOTE: because this event is a result of batch containing GRPC_OP_RECV_STATUS_ON_CLIENT,
            // success will be always set to true.

            using (Profilers.ForCurrentThread().NewScope("AsyncCall.HandleUnaryResponse"))
            {
                TResponse msg = default(TResponse);
                var deserializeException = TryDeserialize(receivedMessage, out msg);

                lock (myLock)
                {
                    finished = true;

                    if (deserializeException != null && receivedStatus.Status.StatusCode == StatusCode.OK)
                    {
                        receivedStatus = new ClientSideStatus(DeserializeResponseFailureStatus, receivedStatus.Trailers);
                    }
                    finishedStatus = receivedStatus;

                    ReleaseResourcesIfPossible();
                }

                responseHeadersTcs.SetResult(responseHeaders);

                var status = receivedStatus.Status;

                if (status.StatusCode != StatusCode.OK)
                {
                    unaryResponseTcs.SetException(new RpcException(status));
                    return;
                }

                unaryResponseTcs.SetResult(msg);
            }
        }`

**Comment by jtattermusch** [comment]
Yes, you are right that for the blocking unary calls, there's no API to access the trailers. The blocking overload is meant as a convenience and you can always use the non-blocking version of unary call if you need access to the full API. Is using the non-blocking version a problem for you?

One problem with giving access to trailers through RpcException is that the protocol allows you to also send trailers with an OK status (in which case you don't receive RpcException and there is nowhere to attach trailers besides the existing API for accessing metadata).
That said, adding trailers field to the RpcException class seems doable and we might consider that for some of the future releases (right now the biggest priority is to deliver CoreCLR support).

**Comment by vjpai** [comment]
Please comment on whether or not this bug should remain open.

**Comment by apolcyn** [comment]
I think this can stay open to keep it on the road map

**Comment by jtattermusch** [comment]
I am planning no action here in the near future.

**Comment by jtattermusch** [comment]
FTR @neuecc would like to have this feature.

**Comment by neuecc** [comment]
@jtattermusch 

I've commented why RpcException. https://github.com/grpc/grpc/issues/9235#issuecomment-273052274
Current gRPC can not put Japanese in Status.Detail, so as a workaround, we are going to put a message on Trailers instead.
This is a common process to all calls.

Ideally I would like to implement such an implementation.

```csharp
public static class RpcExceptionExtensions
{
    public static string GetStatusDetailFromTrailers(this RpcException ex)
    {
        // Status.Detail can't take japanese : https://github.com/grpc/grpc/issues/9235
        return ex.GetTrailers()["error.details"].Value;
    }
}
```

Here is a class probably we will make.

```csharp
public interface IResponseAsyncIntercepter
{
    Task<T> ResponseAsync<T>(Func<Task<T>> responseAsync);
}

// inject this to invoker
public class ShowDialogIntercepter
{
    public async Task<T> ResponseAsync<T>(Func<Task<T>> responseAsync)
    {
        try
        {
            return await responseAsync();
        }
        catch (RpcException ex)
        {
            var errorDetail = ex.GetStatusDetailFromTrailers();
            MyDialog.Show(errorDetail); // show to user.
            throw new TaskCanceledException("top level code handle this and ignore.", ex);
        }
    }
}
```

However, in this case, AsyncUnaryCall can be passed, so I think that it can be avoided as follows.

```csharp
Task<T> ResponseAsync<T>(Func<Metadata> getTrailers, Func<Task<T>> responseAsync);
```

**Comment by jtattermusch** [comment]
I am not opposed to reconsidering adding the trailers to RpcException,
but  the existence of #9235 is a poor reason to do so (there might be others). The right thing is to fix #9235 (I have a fix ready, but it depends on other work to be merged). Fixing #9235 is actually simpler than the workarounds you are suggesting (see the fix in https://github.com/grpc/grpc/pull/9390).

**Comment by schmitch** [comment]
Why is the existence of #9235 a poor reason to support trailing headers in RpcException?

I mean consider storing Bytes as Metadata, there is no reason to store them inside the Data/Description field.
Consider adding a Protobuf to the Metadata to give a more detailed version of an error, no reason to do that inside a String field?

**Comment by jtattermusch** [comment]
@schmitch   #9235 is a bug in encoding of status details string, and adding a new API element to workaround a bug that can easily be fixed (and already was fixed) is a bad idea.

And yes, you can store binary data in an RPC trailer and use it as protobuf-encoded details about an error (see https://github.com/grpc/grpc/pull/11619/files for an example how to do that), but that has nothing to do with #9235.

**Comment by apolcyn** [comment]
should this stay open?

**Comment by jtattermusch** [comment]
Fixed by  #12120.
